dds/DCPS/XTypes/IdlScanner.h

```diff
@@ -0,0 +1,890 @@
+/*
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#ifndef OPENDDS_DCPS_XTYPES_IDL_SCANNER_H
+#define OPENDDS_DCPS_XTYPES_IDL_SCANNER_H
+
+#ifndef OPENDDS_SAFETY_PROFILE
+
+#  include <dds/DCPS/dcps_export.h>
+
+#  include <dds/DCPS/Definitions.h>
+#  include <dds/DdsDynamicDataC.h>
+
+#  include <string>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace XTypes {
+
+class OpenDDS_Dcps_Export CharacterScanner {
+public:
+  CharacterScanner(const std::string& str)
+    : str_(str)
+    , idx_(0)
+  {}
+
+  char peek() const
+  {
+    OPENDDS_ASSERT(!eoi());
+    return str_[idx_];
+  }
+
+  bool eoi() const
+  {
+    return idx_ == str_.size();
+  }
+
+  void consume()
+  {
+    OPENDDS_ASSERT(!eoi());
+    ++idx_;
+  }
+
+  bool match(char c)
+  {
+    if (eoi() || peek() != c) {
+      return false;
+    }
+
+    consume();
+    return true;
+  }
+
+private:
+  const std::string str_;
+  size_t idx_;
+};
+
+class OpenDDS_Dcps_Export IdlToken {
+public:
+  enum Kind {
+    Error,
+    EOI,
+    BooleanLiteral,
+    IntegerLiteral,
+    FloatLiteral,
+    StringLiteral,
+    CharacterLiteral,
+    Identifier
```

Comment:
Identifiers can appear for boolean literals (`TRUE` or `FALSE`) and for enum values.  Section 7.5.2.7.3 shows an illegal `default_value` based on an enum.  While the example is illegal, it suggests that enum values can be treated like literals.

Reply:
Got it now. They sound like `Constant` to me than `Identifier` but I'm fine with it.

Reply:
If I were to implement a parser for IDL, I would populate the symbol table with constants for `TRUE` and `FALSE` so that the identifier would return the appropriate constant.